web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts

```diff
@@ -54,42 +60,219 @@ export class AccessRequestsService {
-      Object.keys(pendingAccessRequest.user_group).length
-    );
+    if (pendingAccessRequest.kind === 'role') {
+      return pendingAccessRequest.roles.size;
+    }
+    if (pendingAccessRequest.kind === 'resource') {
+      return pendingAccessRequest.resources.size;
+    }
+
+    return 0;
+  }
+
+  async addOrRemoveResource({
+    kind,
+    resource,
+  }: ResourceRequest): Promise<void> {
+    this.setState(draftState => {
+      if (draftState.pending.kind !== 'resource') {
+        draftState.pending = {
+          kind: 'resource',
+          resources: new Map(),
+        };
+      }
+
+      const { resources } = draftState.pending;
+
+      if (resources.has(resource.uri)) {
+        resources.delete(resource.uri);
+      } else {
+        // Store only properties required by the type.
+        if (kind === 'app' || kind === 'database' || kind === 'kube') {
+          resources.set(resource.uri, {
+            kind,
+            resource: { uri: resource.uri },
+          });
+        } else {
+          resources.set(resource.uri, {
+            kind,
+            resource: { uri: resource.uri, hostname: resource.hostname },
+          });
+        }
+      }
+    });
   }
 
-  addOrRemoveResource(
-    kind: ResourceKind,
-    resourceId: string,
-    resourceName: string
-  ) {
+  async addOrRemoveRole(role: string): Promise<void> {
     this.setState(draftState => {
-      const kindIds = draftState.pending[kind];
-      if (kindIds[resourceId]) {
-        delete kindIds[resourceId];
+      if (draftState.pending.kind !== 'role') {
+        draftState.pending = {
+          kind: 'role',
+          roles: new Set(),
+        };
+      }
+
+      const { roles } = draftState.pending;
+      if (roles.has(role)) {
+        roles.delete(role);
       } else {
-        kindIds[resourceId] = resourceName || resourceId;
+        roles.add(role);
       }
     });
   }
 }
 
-export function getEmptyPendingAccessRequest() {
+/** Returns an empty access request. We default to the resource access request. */
+export function getEmptyPendingAccessRequest(): PendingAccessRequest {
   return {
-    node: {},
-    db: {},
-    kube_cluster: {},
-    app: {},
-    role: {},
-    windows_desktop: {},
-    user_group: {},
+    kind: 'resource',
+    resources: new Map(),
   };
 }
+
+export type PendingAccessRequest =
+  | {
+      kind: 'resource';
+      resources: Map<ResourceUri, ResourceRequest>;
+    }
+  | { kind: 'role'; roles: Set<string> };
+
+export type ResourceRequest =
+  | {
+      kind: 'server';
+      resource: {
+        uri: ServerUri;
+        hostname: string;
+      };
+    }
+  | {
+      kind: 'database';
+      resource: {
+        uri: DatabaseUri;
+      };
+    }
+  | {
+      kind: 'kube';
+      resource: {
+        uri: KubeUri;
+      };
+    }
+  | {
+      kind: 'app';
+      resource: {
+        uri: AppUri;
+      };
+    };
+
+type SharedResourceAccessRequestKind = 'app' | 'db' | 'node' | 'kube_cluster';
```

Comment:
I wasn't sure if it's best to redefine such type here or define it by using a type from the shared package (well, the teleport package in this case). But in the end I guess it doesn't matter much as those are just simple strings and TypeScript will tell us if something doesn't match.

And if we wanted to use a type from the shared package, we'd have to do something like `Exclude<>` or `Extract<>` anyway.

Reply:
Yeah, I think re-declared type here is easier to read than `Exclude` (I'm also not sure what is the source of truth for this union). 
But we will get a TS error if the types don't match anyway.